modules/ROOT/pages/extending-neo4j/procedures.adoc

```diff
@@ -190,9 +190,94 @@ The classes that can be injected are:
 * `Transaction`
 //* `SecurityContext`
 //* `ProcedureTransaction`
+//* `ProcedureMemory` Candidate for public API but not stable yet
 
 All of the above classes are considered safe and future-proof and do not compromise the security of the database.
 Several unsupported (restricted) classes can also be injected and can be changed with little or no notice.
 Procedures written to use these restricted APIs are not loaded by default, and you need to use the `dbms.security.procedures.unrestricted` to load unsafe procedures.
 Read more about this config setting in link:{neo4j-docs-base-uri}/operations-manual/{page-version}/security/securing-extensions[Operations Manual -> Securing extensions].
 
+[[memory-resource-tracking]]
+== Memory Resource Tracking
+
+[NOTE]
+====
+The memory resource tracking API for procedure framework is available for preview.
+Future versions of neo4j might contain breaking changes to this API.
+====
+
+If your procedure allocates significant amounts of heap memory,
+you can register allocations to count towards the
+link:{neo4j-docs-base-uri}/operations-manual/{page-version}/performance/memory-configuration/#memory-configuration-limit-transaction-memory[configured transaction limits].
+This allows you to avoid `OutOfMemory` errors that cause database restarts.
+Memory allocations will also show up in query profiles.
+
+To do this you need to inject a `org.neo4j.procedure.memory.ProcedureMemory` as a field in your procedure class.
+`ProcedureMemory` has various methods to allow you to register allocations.
+For example (see javadocs for a full reference):
+
+* `ProcedureMemoryTracker newTracker()`,
+  creates a new memory resource tracker that is bound to the current transaction.
+* `HeapEstimator heapEstimator()`,
+  used to estimate the heap size of classes and instances.
+* `HeapTrackingCollectionFactory collections()`,
+  lets you create collections that have build-in memory tracking of their internal structure.
+
+It's usually difficult and time-consuming to implement memory resource tracking.
+Here's a few considerations and caveats that are worth keeping in mind:
+
+- Limit the scope of the memory management.
+  Focus only on parts that can grow significantly in memory, and ignore minor underestimation.
+- Beware of overestimation by registering allocations of the same instance multiple times.
+  It's up to you to add reference counting or other mechanisms to avoid this if it's an issue for you.
+- It's common to not know the size of an instance before it has been allocated.
+  This makes you register allocations after they have already been made.
+  The memory tracker implementation try to guard against this
+  by always pre-registering a certain amount of memory in the internal memory pools.
```

Comment:
```suggestion
  The memory tracker implementation tries to prevent this by always pre-registering a certain amount of memory in the internal memory pools.
```